### PR TITLE
TCPIP Section Updates

### PR DESCRIPTION
**Pull request by nibanks**

### CODE REVIEW

File: udp-rsc-offload.md [hunk 1]
```diff
@@ -19,15 +19,14 @@ The Winsock API (currently only software fallback) already exists, and details o
...
-- If hardware URO is enabled, but a socket opts-in to a smaller max coalesced size, TCPIP/AFD will break the coalesced receive into the smaller size for the socket.
-- If hardware URO is enabled, but a socket does not opt-in to URO, then TCPIP will resegment receives for that socket.
-- If hardware URO is not available, but a socket opts-in to URO with the Winsock API, software URO will be used for that socket. Software coalescing will follow the same rules as NDIS, except for the following:
-    - Software coalescing may decide to use multiple NBLs in a chain of up to 255, instead of a single NBL.
-    - Software coalescing will not coalesce IP fragments, broadcast or multicast datagrams, or datagrams with IP extension headers.
-- TCPIP will disable the hardware offload, and the software fallback, if any WFP/LWF filters that are URO-incompatible are installed on the system.
+This section describes necessary updates in the Windows network stack (TCPIP & AFD) when hardware URO support is enabled:
+
+- If a socket opts-in to URO with a max coalesced size greater than or equal to the hardare offload size, then the OS will deliver the NBLs from hardware unmodified to the socket.
```
Comment (mtfriesen) [inline udp-rsc-offload.md:24]:
```suggestion
- If a socket opts-in to URO with a max coalesced size greater than or equal to the hardware offload size, then the OS will deliver the NBLs from hardware unmodified to the socket.
```

File: udp-rsc-offload.md [hunk 2]
```diff
@@ -19,15 +19,14 @@ The Winsock API (currently only software fallback) already exists, and details o
...
+- If a socket opts-in to a smaller max coalesced size, the OS will break the coalesced receive into the smaller size for the socket.
+- If a socket does not opt-in to URO, then the OS will resegment receives for that socket.
+- If an incompatible WFP filter is encountered, then the OS will resegment receives for that filter.
+
+In the absense of hardware URO, the existing software URO feature will continue to be available.
```
Comment (mtfriesen) [inline udp-rsc-offload.md:29]:
```suggestion
In the absence of hardware URO, the existing software URO feature will continue to be available.
```

File: udp-rsc-offload.md [hunk 3]
```diff
@@ -19,15 +19,14 @@ The Winsock API (currently only software fallback) already exists, and details o
...
-- If hardware URO is enabled, but a socket opts-in to a smaller max coalesced size, TCPIP/AFD will break the coalesced receive into the smaller size for the socket.
-- If hardware URO is enabled, but a socket does not opt-in to URO, then TCPIP will resegment receives for that socket.
-- If hardware URO is not available, but a socket opts-in to URO with the Winsock API, software URO will be used for that socket. Software coalescing will follow the same rules as NDIS, except for the following:
-    - Software coalescing may decide to use multiple NBLs in a chain of up to 255, instead of a single NBL.
-    - Software coalescing will not coalesce IP fragments, broadcast or multicast datagrams, or datagrams with IP extension headers.
-- TCPIP will disable the hardware offload, and the software fallback, if any WFP/LWF filters that are URO-incompatible are installed on the system.
+This section describes necessary updates in the Windows network stack (TCPIP & AFD) when hardware URO support is enabled:
```
Comment (mtfriesen) [inline udp-rsc-offload.md:22]: Should we add a note for forwarding packets (re-segment URO if necessary, or convert to USO)? What about stuff like IPSNPI?

Comment (nibanks) [reply udp-rsc-offload.md:22]: Yeah, probably. What do you suggest. Do you just want to merge this and make a new PR to add stuff for that?

Comment (mtfriesen) [reply udp-rsc-offload.md:22]: I'm fine with merging this as-is, as the list of changes to TCPIP is likely to grow substantially as we discover more issues.

The other question is: how much of this do we need to publicly document, and for the truly implementation details side of things, should we track internally?

**Pull request merged**